### PR TITLE
✨ Contribute experimental cluster-api-state-metrics

### PR DESCRIPTION
**Pull request by tobiasgiese**
Signed-off-by: Tobias Giese <tobias.giese@mercedes-benz.com>
Co-authored-by: Christian Schlotter <schlotterc@vmware.com>

**What this PR does / why we need it**:

This PR contributes the [Cluster-API State Metrics](https://github.com/mercedes-benz/cluster-api-state-metrics/) exporter to the Cluster-API core repository.
As this is an experimental feature -- as discussed in the design proposal [cluster-api-state-metrics](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220411-cluster-api-state-metrics.md) -- we will store the code in `exp/`.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #6458

<sub>Tobias Giese <tobias.giese@mercedes-benz.com>, Mercedes-Benz Tech Innovation GmbH, [legal info/Impressum](https://github.com/mercedes-benz/foss/blob/master/LEGAL_IMPRINT.md)</sub>

### CODE REVIEW

File: exp/state-metrics/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# syntax=docker/dockerfile:1.1-experimental
```
Comment (chrischdi) [inline exp/state-metrics/Dockerfile:1]: We can remove this :-)

File: Makefile [hunk 2]
```diff
@@ -509,6 +526,12 @@ docker-build-kubeadm-control-plane: ## Build the docker image for kubeadm contro
 	$(MAKE) set-manifest-image MANIFEST_IMG=$(KUBEADM_CONTROL_PLANE_CONTROLLER_IMG)-$(ARCH) MANIFEST_TAG=$(TAG) TARGET_RESOURCE="./controlplane/kubeadm/config/default/manager_image_patch.yaml"
 	$(MAKE) set-manifest-pull-policy TARGET_RESOURCE="./controlplane/kubeadm/config/default/manager_pull_policy.yaml"
 
+.PHONY: docker-build-cluster-api-state-metrics
+docker-build-cluster-api-state-metrics: ## Build the docker image for state metrics
+	cd $(STATE_METRICS_DIR); DOCKER_BUILDKIT=1 docker build --build-arg builder_image=$(GO_CONTAINER_IMAGE) --build-arg goproxy=$(GOPROXY) --build-arg ARCH=$(ARCH) --build-arg ldflags="$(LDFLAGS)" . -t $(STATE_METRICS_IMG)-$(ARCH):$(TAG)
```
Comment (chrischdi) [inline Makefile:531]: This way we can get rid of the copied Dockerfile :-)

```suggestion
	cd $(STATE_METRICS_DIR); DOCKER_BUILDKIT=1 docker build --build-arg builder_image=$(GO_CONTAINER_IMAGE) --build-arg goproxy=$(GOPROXY) --build-arg ARCH=$(ARCH) --build-arg ldflags="$(LDFLAGS)" . -t $(STATE_METRICS_IMG)-$(ARCH):$(TAG) -f ../../Dockerfile
```

Comment (chrischdi) [inline Makefile:531]: This way we can get rid of the copied Dockerfile :-)

```suggestion
	cd $(STATE_METRICS_DIR); DOCKER_BUILDKIT=1 docker build --build-arg builder_image=$(GO_CONTAINER_IMAGE) --build-arg goproxy=$(GOPROXY) --build-arg ARCH=$(ARCH) --build-arg ldflags="$(LDFLAGS)" . -t $(STATE_METRICS_IMG)-$(ARCH):$(TAG) -f ../../Dockerfile
```

File: exp/state-metrics/config/default/kustomization.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+- ../rbac
+- ../manager
+
+patchesStrategicMerge:
+- manager_pull_policy.yaml
+
+images:
+- name: controller
+  newName: docker.pkg.github.com/daimler/cluster-api-state-metrics/cluster-api-state-metrics
+  newTag: latest
```
Comment (chrischdi) [inline exp/state-metrics/config/default/kustomization.yaml:32]:
```suggestion
```

Comment (chrischdi) [inline exp/state-metrics/config/default/kustomization.yaml:32]:
```suggestion

```

**Review by chrischdi — COMMENTED** [review]
Some small nits.


Many thanks for the contribution :-)

For everyone joining the party here: This PR contains the contribution from the mercedes-benz folks. #6458 outlines the follow-up steps like e.g. the implementation of metrics missing currently but proposed in the proposal, tilt integration, etc.

File: exp/state-metrics/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+ARG package=.
+ARG ARCH
+ARG ldflags
+
+# Do not force rebuild of up-to-date packages (do not use -a) and use the compiler cache folder
+RUN --mount=type=cache,target=/root/.cache/go-build \
+    --mount=type=cache,target=/go/pkg/mod \
+    CGO_ENABLED=0 GOOS=linux GOARCH=${ARCH} \
+    go build -trimpath -ldflags "${ldflags} -extldflags '-static'" \
+    -o manager ${package}
```
Comment (chrischdi) [inline exp/state-metrics/Dockerfile:55]:
```suggestion
    cd ${package} && \
    CGO_ENABLED=0 GOOS=linux GOARCH=${ARCH} \
    go build -trimpath -ldflags "${ldflags} -extldflags '-static'" \
    -o /workspace/manager .
```

File: Makefile (hunk 2 above)
Comment (chrischdi) [inline Makefile:531]:
```suggestion
	DOCKER_BUILDKIT=1 docker build --build-arg builder_image=$(GO_CONTAINER_IMAGE) --build-arg goproxy=$(GOPROXY) --build-arg ARCH=$(ARCH) --build-arg package=./exp/state-metrics --build-arg ldflags="$(LDFLAGS)" . -t $(STATE_METRICS_IMG)-$(ARCH):$(TAG) -f $(STATE_METRICS_DIR)/Dockerfile
```

**Review by chrischdi — COMMENTED** [review]
Fixes for the replace

**Comment by chrischdi** [comment]
I think we are nearly ready to squash + remove WIP :-)

File: Makefile [hunk 5]
```diff
@@ -244,6 +249,13 @@ generate-manifests-kubeadm-control-plane: $(CONTROLLER_GEN) ## Generate manifest
 		output:webhook:dir=./controlplane/kubeadm/config/webhook \
 		webhook
 
+.PHONY: generate-manifests-cluster-api-state-metrics
+generate-manifests-cluster-api-state-metrics: $(CONTROLLER_GEN) ## Generate RBAC cluster-api-state-metrics
```
[the remark this answers is not in the document]
Comment (tobiasgiese) [reply Makefile:253]: Or we could add `cluster-api-state-metrics` to l.194. WDYT?

**Comment by chrischdi** [comment]
Thanks again for the intent to contribute this 👍 . Because https://github.com/kubernetes/kube-state-metrics/pull/1777  got merged, we will move forward using the Custom Resource Configuration.

The code here was still helpful because it provided some first example metrics 🎉 

/close

**Pull request closed without merging**